File: src/main/java/dev/foxgirl/mineseekdestroy/util/Region.java
package dev.foxgirl.mineseekdestroy.util;

import kotlin.jvm.internal.ArrayIteratorKt;
import net.minecraft.util.math.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;

public final class Region implements Selection {

    public static final @NotNull Region EMPTY = new Region(
        new BlockPos(0, -64, 0),
        new BlockPos(0, -64, 0)
    );

    private final BlockPos start;
    private final BlockPos end;
    private final ChunkPos chunkStart;
    private final ChunkPos chunkEnd;
    private final Box box;

    public Region(@NotNull Vec3i a, @NotNull Vec3i b) {
        Objects.requireNonNull(a, "Argument 'a'");
        Objects.requireNonNull(b, "Argument 'b'");
        start = new BlockPos(
            Math.min(a.getX(), b.getX()),
            Math.min(a.getY(), b.getY()),
            Math.min(a.getZ(), b.getZ())
        );
        end = new BlockPos(
            Math.max(a.getX(), b.getX()),
            Math.max(a.getY(), b.getY()),
            Math.max(a.getZ(), b.getZ())
        );
        chunkStart = new ChunkPos(start);
        chunkEnd = new ChunkPos(end);
        box = Box.enclosing(start, end.add(1, 1, 1));
    }

    public @NotNull BlockPos component1() {
        return start;
    }
    public @NotNull BlockPos component2() {
        return end;
    }

    public @NotNull BlockPos getStart() {
        return start;
    }
    public @NotNull BlockPos getEnd() {
        return end;
    }
    public @NotNull ChunkPos getChunkStart() {
        return chunkStart;
    }
    public @NotNull ChunkPos getChunkEnd() {
        return chunkEnd;
    }
    public @NotNull Box getBox() {
        return box;
    }

    public @NotNull Vec3d getCenter() {
        return new Vec3d(
            (double) ((long) start.getX() + (long) end.getX()) / 2.0 + 0.5,
            (double) ((long) start.getY() + (long) end.getY()) / 2.0 + 0.5,
            (double) ((long) start.getZ() + (long) end.getZ()) / 2.0 + 0.5
        );
    }

    public long getBlockCount() {
        return
            (long) (end.getX() - start.getX() + 1) *
            (long) (end.getY() - start.getY() + 1) *
            (long) (end.getZ() - start.getZ() + 1);
    }
    public long getChunkCount() {
        return
            (long) (chunkEnd.x - chunkStart.x + 1) *
            (long) (chunkEnd.z - chunkStart.z + 1);
    }

    @Override
    public boolean contains(int x, int y, int z) {
        return start.getX() <= x && end.getX() >= x
            && start.getY() <= y && end.getY() >= y
            && start.getZ() <= z && end.getZ() >= z;
    }

    public @NotNull Region offset(int x, int y, int z) {
        return new Region(start.add(x, y, z), end.add(x, y, z));
    }
    public @NotNull Region offset(@NotNull Vec3i pos) {
        return offset(pos.getX(), pos.getY(), pos.getZ());
    }

    public boolean equals(@Nullable Object other) {
        if (other == this) return true;
        return other instanceof Region otherRegion
            && start.equals(otherRegion.start)
            && end.equals(otherRegion.end);
    }

    public int hashCode() {
        return 31 * start.hashCode() + end.hashCode();
    }
    public @NotNull String toString() {
        return "Region{start=" + start + ", end=" + end + "}";
    }

    public static final class Set extends AbstractSet<Region> implements Selection {

        private final Region[] regions;

        public Set(@NotNull Region... elements) {
            this(Arrays.asList(elements));
        }
        public Set(@NotNull Collection<Region> collection) {
            regions = new LinkedHashSet<>(collection).toArray(new Region[0]);
        }

        @Override
        public boolean contains(int x, int y, int z) {
            for (var region : regions) if (region.contains(x, y, z)) return true;
            return false;
        }

        @Override
        public int size() {
            return regions.length;
        }
        @Override
        public @NotNull Iterator<Region> iterator() {
            return ArrayIteratorKt.iterator(regions);
        }

    }

}

File: src/main/java/dev/foxgirl/mineseekdestroy/util/Region.kt
package dev.foxgirl.mineseekdestroy.util

import dev.foxgirl.mineseekdestroy.util.collect.ImmutableSet
import net.minecraft.util.math.*

class Region : Selection {

    val start: BlockPos
    val end: BlockPos

    val chunkStart: ChunkPos
    val chunkEnd: ChunkPos

    val box: Box

    constructor(a: Vec3i, b: Vec3i) {
        start = BlockPos(
            Math.min(a.x, b.x),
            Math.min(a.y, b.y),
            Math.min(a.z, b.z),
        )
        end = BlockPos(
            Math.max(a.x, b.x),
            Math.max(a.y, b.y),
            Math.max(a.z, b.z),
        )
        chunkStart = ChunkPos(start)
        chunkEnd = ChunkPos(end)
        box = Box(start, end.add(1, 1, 1))
    }

    val center get() =
        Vec3d(
            (start.x.toLong() + end.x.toLong()).toDouble() / 2.0 + 0.5,
            (start.y.toLong() + end.y.toLong()).toDouble() / 2.0 + 0.5,
            (start.z.toLong() + end.z.toLong()).toDouble() / 2.0 + 0.5,
        )

    val blockCount get() =
        (end.x - start.x + 1).toLong() *
        (end.y - start.y + 1).toLong() *
        (end.z - start.z + 1).toLong()

    val chunkCount get() =
        (chunkEnd.x - chunkStart.x + 1).toLong() *
        (chunkEnd.z - chunkStart.z + 1).toLong()

    operator fun component1() = start
    operator fun component2() = end

    override fun contains(x: Int, y: Int, z: Int) =
        start.x <= x && end.x >= x &&
        start.y <= y && end.y >= y &&
        start.z <= z && end.z >= z

    fun offset(x: Int, y: Int, z: Int) = Region(start.add(x, y, z), end.add(x, y, z))
    fun offset(pos: Vec3i) = offset(pos.x, pos.y, pos.z)

    override fun equals(other: Any?) =
        other === this || (other is Region && start == other.start && end == other.end)

    override fun hashCode() =
        31 * start.hashCode() + end.hashCode()

    override fun toString() =
        "Region{start=$start, end=$end}"

    class Set : AbstractSet<Region>, Selection {

        private val regions: Array<Region>

        constructor(vararg elements: Region) {
            regions = ImmutableSet.copyOf(elements).toTypedArray()
        }
        constructor(collection: Collection<Region>) {
            regions = ImmutableSet.copyOf(collection).toTypedArray()
        }

        override fun contains(x: Int, y: Int, z: Int) = regions.any { it.contains(x, y, z) }

        override val size get() = regions.size
        override fun iterator() = regions.iterator()

    }

}
